File: DSA/DSA02005 HOÁN VỊ XÂU KÝ TỰ.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
void recursion(vector<int> store, vector<bool> placed, int j)
{
    for (int i = 0; i < n; i++)
    {
        if (!placed[i])
        {
            store[j] = i;
            placed[i] = true;

            if (j == n-1)
            {
                for (auto &&k : store)
                {
                    cout << char(k+'A');
                }
                cout << " ";
            }
            else
            {
                recursion(store, placed, j+1);
            }
            placed[i] = false;
        }
    }
}

void submit()
{
    string A;
    cin >> A;
    n = A.size();
    vector<int> store(n);
    vector<bool> placed(n, false);

    recursion(store, placed, 0);
    cout << endl;
}   

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();

    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}

File: DSA_python/DSA02005 HOÁN VỊ XÂU KÝ TỰ.py
def recursion(store, placed, j):
    for i in range(n):
        if placed[i] == False:
            store[j] = i
            placed[i] = True

            if j == n-1:
                for i in store:
                    print(chr(i+ord("A")), end="")
                print(end=" ")
            else:
                recursion(store, placed, j+1)
            placed[i] = False

def submit():
    global n
    A = str(input())
    n = len(A)
    store = [0]*n
    placed = [False]*n

    recursion(store, placed, 0)
    print()

test = int(input())
for i in range(test):
    submit()